connector/src/main/java/org.apache.flink/connector/nebula/sink/NebulaBatchOutputFormat.java

```diff
@@ -116,49 +105,109 @@ public void open(int i, int i1) throws IOException {
+                    graphProvider.getPassword(), true);
+        } catch (NotValidConnectionException | AuthFailedException
+                 | ClientServerIncompatibleException | IOErrorException e) {
+            LOG.error("failed to get graph session", e);
+            throw new IOException("get graph session error", e);
+        }
+        ResultSet resultSet;
+        try {
+            resultSet = session.execute("USE " + executionOptions.getGraphSpace());
+        } catch (IOErrorException e) {
+            LOG.error("switch space error", e);
+            throw new IOException("switch space error", e);
+        }
+        if (!resultSet.isSucceeded()) {
+            LOG.error("switch space failed: " + resultSet.getErrorMessage());
+            throw new IOException("switch space failed: " + resultSet.getErrorMessage());
+        }
+    }
+
     protected abstract NebulaBatchExecutor<T> createNebulaBatchExecutor();
 
     /**
      * write one record to buffer
      */
     @Override
-    public final synchronized void writeRecord(T row) {
+    public final synchronized void writeRecord(T row) throws IOException {
+        checkCommitException();
         nebulaBatchExecutor.addToBatch(row);
+        numPendingRow++;
 
-        if (numPendingRow.incrementAndGet() >= executionOptions.getBatchSize()) {
+        if (executionOptions.getBatchSize() > 0
+                && numPendingRow >= executionOptions.getBatchSize()) {
             commit();
         }
     }
 
     /**
      * commit batch insert statements
      */
-    private synchronized void commit() {
-        String errorExec = nebulaBatchExecutor.executeBatch(session);
-        if (errorExec != null) {
-            errorBuffer.add(errorExec);
+    private synchronized void commit() throws IOException {
+        int maxRetries = executionOptions.getMaxRetries();
+        int retryDelayMs = executionOptions.getRetryDelayMs();
+        boolean failOnError = executionOptions.getFailureHandler().equals(FailureHandlerEnum.FAIL);
+
+        // execute the batch at most `maxRetries + 1` times
+        for (int i = 0; i <= maxRetries; i++) {
+            try {
+                nebulaBatchExecutor.executeBatch(session);
+                numPendingRow = 0;
+                break;
+            } catch (Exception e) {
+                LOG.warn(String.format("write data error (attempt %d)", i), e);
+                boolean nonRecoverable = (e instanceof NebulaBatchExecutor.ExecutionException)
+                        && ((NebulaBatchExecutor.ExecutionException) e).isNonRecoverableError();
+                if (i >= maxRetries || nonRecoverable) {
+                    // clear the batch on failure when we do not want more retries
+                    nebulaBatchExecutor.clearBatch();
+                    numPendingRow = 0;
+                    if (failOnError) {
+                        throw e;
+                    }
+                    renewSession();
```

Comment:
I think the only condition to renewSession is  there are SESSION_INVALD and SESSION_TIMEOUT error code. 
* For IOException, we do not need to renew `Session` but need to renew Session's `Connection`, then we can reduce the re-auth opertion.
* For EXECUTION_ERROR error code, the reason is on Storaged, we can just re-execute with the same `Session`.

Reply:
@Nicole00 Yeah, I can see that there may be situations where renewing the session is unnecessary, but here I took a simple approach to make sure the subsequent operation always starts in a clean state.

* For `IOException`, right now we use the `NebulaPool.getSession()` method to get the session. I looked at its implementation and it internally manages a pool of connections. There may be opportunities to avoid unnecessary authentication, but that may require us to manage sessions/connections in the Flink connector itself. This is probably a bigger body of work. Maybe this is something we can consider in the future, or something that can be improved inside the NebulaGraph Java client?
* For `E_EXECUTION_ERROR` and many other error codes, I'm worried about the situation of "partial success" even the final outcome is an error. For example, there may be a temporary issue with one of the storaged services while the other ones function normally. I've also seen the session times out while the slow query is still being executed in storaged. This is problematic in Flink applications where we process a mix of insert/update/delete events, and concurrent writes to the same key may result in out-of-order processing and corrupted data at the sink. Therefore, I'd like to renew the session anyway, which proactively kills the existing session (and any query that are still running), so that we *rule out the possibility of concurrent queries*. In other words, in Flink I'd prefer there is only one query running for the session attached to each Flink task, and I'd be extremely careful about session reuse after an error.

In general, I think the current "catch-all" session renew logic does not affect the correctness of error handling, and it's robust due to its simplicity. There may be situations where some renewals can be avoided, but we could treat this as an optimization to revisit in some future PR. What do you think?

Reply:
@linhr Yeah, you are right, `renew session anyway` is totally a correct and simplest way when query error,  and is also the roughest way 😁. 
*  java client indeed has retry  mechanism for IOException, need to config the `reconnect` as true when `getSession`, so we do not need to process this in Flink connector.
* And For the concurrent query you mentioned, it does not exist here, because the session does not support concurrent query, and the client has a synchronized lock on `execute` interface.
* At last, I think it's ok to optimize the re-execute logic in future, and maybe we will replace the java client session to `SessionPool`, which already process the re-execute and we just need to provide some configs.

Reply:
@Nicole00 Thanks for the comment! Yeah it would be great to further improve the code in the future using the built-in error handling support in the Java client.

Regarding the concurrent query, let me try to explain the situation a bit more. If the previous query fails, it may not have failed cleanly. For example, when the query times out, we get the error and the lock is released. At this point we can run another query to insert the same data using the same session. However, the previous failed query is still being executed, so we get concurrent queries. Since we do not have transactional writes and the previous execution is not rolled back, the actual data finally written is not well defined. This is of course an edge case, but I was concerned that it may lead to corrupted data, so I renewed the session to be safe.

Happy to discuss!